### DISCUSSION TITLE
Serialization with JSON works but fails with YAML

### DISCUSSION THREAD

**Discussion by m-31** [charleskorn/kaml#189] [Q&A] [answered]
I have a problem with custom serialization when using YAML. In the example below I want to serialize a <Map<String, String>> object like a <List<Map<String, String>>> object. With JSON this works flawlessly. With YAML I get an Exception that looks like this:

    Exception in thread "main" com.charleskorn.kaml.InvalidPropertyValueException at parameters on line 2, column 1: Value for 'parameters' is invalid: Expected a map, but got a list
	at com.charleskorn.kaml.YamlObjectInput.decodeElementIndex(YamlInput.kt:363)

Do you have any ideas why this fails?

This is the code I used:

        object MapAsListMapSerializer : KSerializer<Map<String, String>> {
            override val descriptor: SerialDescriptor = MapSerializer(String.serializer(), String.serializer()).descriptor
        
            override fun serialize(encoder: Encoder, value: Map<String, String>) {
                encoder.encodeSerializableValue(serializer(), value.map { mapOf(it.key to it.value) })
            }
        
            override fun deserialize(decoder: Decoder): Map<String, String> {
                val list: List<Map<String, String>> = decoder.decodeSerializableValue(ListSerializer(MapSerializer(String.serializer(), String.serializer())))
                return list.map { it.toList().first() }.toMap()
            }
        }
        
        @Serializable
        data class Request(
            @Serializable(with = MapAsListMapSerializer::class)
            val parameters: Map<String, String>
        )
        
        fun main() {
            val requestSerializer: KSerializer<Request> = serializer()
        
            val request = Request(mapOf("i" to "am", "you" to "are"))
    
            val json = Json.encodeToString(requestSerializer, request)
            println(json)
            println(Json.decodeFromString(requestSerializer, json))
        
            val yaml = Yaml().encodeToString(requestSerializer, request)
            println(yaml)
            println(Yaml().decodeFromString(requestSerializer, yaml))
        }

**Comment by charleskorn** [comment, marked as answer]
Hi @m-31, thanks for the question. It looks like the issue might be the fact that `descriptor` in your custom serializer says that the serializer works with maps, but then receives a list. The JSON format is far more accepting of inconsistencies like these, but kaml requires them to be consistent so that it can report more detailed errors.

Something like this should work:

```diff
-override val descriptor: SerialDescriptor = MapSerializer(String.serializer(), String.serializer()).descriptor
+override val descriptor: SerialDescriptor = ListSerializer(MapSerializer(String.serializer(), String.serializer())).descriptor
```

**Reply by m-31** [reply]
Yes, of course, that is the reason. I didn't see the forest for the trees. Thank you very much!
